### ISSUE TITLE
Allowing Pull Requests to the docs

### ISSUE THREAD

**Issue by ChrisMaddock** [nunit/docs#337]
It's been on my mind for a many years that not having a way to pull request our documentation discourages contributors.* Seeing the great contribution in https://github.com/nunit/nunit.analyzers/pull/206 from @SeanKilleen made me think we should really do something about that!

Given the docs have already moved once, I think we want to keep the current wiki pages as their authoritative home. What I'd like to suggest, is cloning the `docs.wiki.git` repo into the main repo of this project, so user's can pull request against that. When a pull request is merged to master, we could then have a CI job which pushes back to `docs.wiki.git`. We'd need to stop editing the wiki directly - I'm not GitHub gives us detailed permissions on wiki-editing, so I think we'd have to just access to a small group of contributors who "know not to do that".

Working this way means we could also potentially push to nunit.org as well in future, if we wanted a prettier interface. ✨ 

Thoughts? 🙂 

---

<details><summary>* To avoid distracting from the issue at hand, my thoughts on that below... 😄</summary>

We originally moved to the wiki to encourage non-technical contributors who might be put off by the git process to contribute. A few years later, those contributors haven't really materialised, and I'd argue that other contributors are put off from making contributions due to:

- No clear path to do so
- Creating additional workload for the small team who can make changes
- Not being able to fully contribute their 'own work'
- Time delay in waiting for a response of how to action a change, or to be give access
- Not getting their 'internet points' and a proper pull request registered in their name 😉 

I'd also say that Microsoft Docs have [shown quite successfully](https://github.com/dotnet/docs/pulls?q=is%3Apr+label%3A%22%3Asparkles%3A+1st-time+docs+contributor%21%22+) that the only GitHub editing interface is pretty accessible, and they also provide some [nice, simple guides](https://docs.microsoft.com/en-gb/contribute/#quick-edits-to-existing-documents) to help with this.

</details>

**Comment by SeanKilleen** [comment]
@ChrisMaddock I'd agree with you (and thanks for the shout out!)

I feel a little ambitious. What about using  the `dotnet/docfx` project to generate a nice docs site with the diffs captured for each version? That might be something I could put together. Then you'd have a nice unified set of docs that can evolve.

**Comment by mikkelbu** [comment]
I don't have a problem with a new location for the documentation (it could e.g. be github pages). Often when I search for something the first hits are the old documentation at https://nunit.org/docs/2.5/testCase.html or the clone of the code at MS https://docs.microsoft.com/en-us/dotnet/api/nunit.framework.testcaseattribute?view=xamarin-ios-sdk-12. So I think a new location that better supports SEO (not that I know much about this) would be apprciated. But I agree that it would be nice to have the same flow as they do on MS docs site (as far as I know not all their tooling is open source, but one can probably find similar functionality). 

I'm unsure about the flow with having a clone that is synced via CI, as I think there are differences between regular markdown and wiki markdown - like links `[[NUnit|NUnit-Documentation]]` vs `[NUnit Gui](https://github.com/CharliePoole/nunit-gui/wiki)`. At least we should only use regular markup and not any of the wiki-markup if we want to go this way.

**Comment by CharliePoole** [comment]
Possibly a minority opinion, but here goes...

We picked the wiki (I did, actually) back in the day because we (1) wanted users to contribute and (2) were willing to give away edit permission to anyone who asked. I think point 2 is what changed.

At one time, I believe I had a a wiki editors team... or maybe I just floated the idea. In any case, it was combined with the Contributors team, which served three purposes...

1. Allowed us to assign issues (no longer needed for that)

2. Allowed editing the wiki

3. Allowed write access to selected repos - primarily samples, IIRC

As a friend and admirer of Ward Cunningham, I always have tried to stick to his notion of the wiki: that interesting things happen when you allow broad access to it. I see direct editing as broad access, while pull requests only give broad access to programmers or others willing to learn how to use Git. The audience for our docs is bigger, I think, than the latter category.

I'd love to see an approach that allows both kinds of access.

**Comment by SeanKilleen** [comment]
@CharliePoole I agree with you 100% in principle; I love the idea of providing the most access to the most people. However, in practice, I've noticed some (admittedly anecdotal) things: 

> were willing to give away edit permission to anyone who asked

I think this can still work with a PR structure; it all depends on the culture that the maintainers adopt. You could even choose to auto-merge pull requests via a bot if more than one person gives a 👍 , for example. (lots of options here).

> pull requests only give broad access to programmers or others willing to learn how to use Git.

I think with GitHub's in-browser editing and preview capabilities, it can be a conceivably lower lift. I also think it's possible that not having a known PR process can lead to self-editing and a lack of contribution. For example, I have seen the NUnit wiki in the past but have never edited it because it didn't seem my place (even if edit access had been given freely to everyone). A PR process tells me there's a known path to make a change in a collaborative way, which allows me to propose something. 

I'd be curious to know if the NUnit docs had actually received any such contributions from folks who would have otherwise been uncomfortable with a PR. My hunch is that there haven't been so many,  given that NUnit is a development library utilized primarily by developers (my calculus would certainly change if I found out it this was not the case). 

Given that, I think it's possible to achieve both aims. I'd suggest:

* Documentation that seamlessly combines docs across all versions (I think `docfx` is a good fit here, though others may work)
* A welcoming pull request process (a template, for example, that welcomes people and walks them through)
* A bot or process that can welcome new contributors
* Thanking contributors in releases
* A bot (if decided) to allow auto-merging by one or more people. 
* Labeling some docs issues/ideas as `help wanted` or `good first issues` or `up for grabs`, etc. and ensuring they're synced to a site like up-for-grabs.net will help establish that NUnit wants to be an on-ramp for folks.

(If you do any of the above already, that's fantastic! I'm not as familiar with the process here yet to know.) 

@CharliePoole do you think that what I outlined above has some promise?

**Comment by ChrisMaddock** [comment]
>  I think there are differences between regular markdown and wiki markdown

Thanks Mikkel - wasn't aware of that.

> (2) were willing to give away edit permission to anyone who asked. I think point 2 is what changed. 

I'd disagree that anything's changed here - simply that nobody is asking for permissions, because a) it's not clear that that's acceptable and b) that puts a 24 hour delay into someone fixing a typo, which I think is often more effort than people are willing to make.

Interestingly, GitHub now has an "anyone can edit the wiki" feature, which I don't think it had at the time? That does feel like a little too open to me however...

In general, I very much agree with Sean's comments above - they reflect my experience of our docs contributors over the last few years.

---

@SeanKilleen On docfx - wow, that _is_ more ambitious than I was planning! 😛 I don't know a lot about the tool, but I'd love for us to have a better docs experience. Can you elaborate a little on just what we could gain with it? Can I throw a few curveballs too? 😅 

- My limited knowledge of docsfx - one of it's big advantage is auto-generating API docs from a codebase, right? You'll see from our docs, we don't currently document each API in the same way e.g. Microsoft do. (It'd be really nice to do for certain parts of certain projects, but a lot of work to set up) Do we lose a lot of the benefit of using it, if we're not documenting by API?
- If we _were_ to start documenting per API - we'd need to work out how we can very specifically limit which APIs are documented, across all the NUnit projects. The engine's a little easier - as there's a specific assembly which is intended to be public, but there's lots of functionality in the framework that is exposed as public, but not intended to be relied on. (Just because "history") 
- Our current docs basically get very little maintenance. That's not the ideal, but reflects our contributors available time! Also, our release processes are already pretty convoluted - if this was going to involve an extra step at release, we'd need to automate as much as possible

I'm only familiar with docfx from Microsoft docs - do you know if there's any projects which are on a more similar level to our docs, currently using it?

**Comment by SeanKilleen** [comment]
@ChrisMaddock I suppose there's [only one way to find out if it's too ambitious! 😁](https://github.com/nunit/nunit/pull/3527)

> One of it's big advantage is auto-generating API docs from a codebase, right? You'll see from our docs, we don't currently document each API in the same way e.g. Microsoft do. (It'd be really nice to do for certain parts of certain projects, but a lot of work to set up) Do we lose a lot of the benefit of using it, if we're not documenting by API?

My understanding is that, if you choose, you get it for the parts you have documented, along with the ability to point to source control. You can choose to set it up that way or not. You can also use articles in the way that the unit docs / wiki do now.

> If we were to start documenting per API - we'd need to work out how we can very specifically limit which APIs are documented, across all the NUnit projects. 

Docfx supports exclusions IIRC; shouldn't be hard to ensure things aren't published if unintended.

> Also, our release processes are already pretty convoluted - if this was going to involve an extra step at release, we'd need to automate as much as possible

The goal would be being able to plop this into your build process (I could assist with that, no problem) and have it be an unobtrusive part of the release.

> do you know if there's any projects which are on a more similar level to our docs, currently using it

[Akka.NET](http://getakka.net) is the first one that comes to mind. I'd plan to use their instance as a reference.

**Comment by ChrisMaddock** [comment]
Nice example - Akka.NET's docs look great!

Personally, I think this would be really nice - although I'm a little concerned about the size of this as a piece of work! The docs in this repo span eight different projects, and that's before we start worrying about documenting APIs - [some of which](https://github.com/nunit/nunit/issues/1426) we can't even work out ourselves if they should be public or not! 😅 

I'd be interested to hear the thoughts of others on the team - @rprouse in particular. 🙂  

(Quick bit of NUnit background - each 'project' within the organisation has it's own maintainer, and then there's also the @nunit/core-team who take oversight of the whole organisation. Rob heads up the framework, which would probably be the most impacted here, and also chairs the Core Team.)

**Comment by SeanKilleen** [comment]
@ChrisMaddock that's really helpful background, thanks! The project seems to have a lot of well-defined groups/structure, and my goal will be to make sure we don't step on too many toes. I want to show what's possible in order to get a go/no-go as soon as we can. I'm also interested in utilizing docfx elsewhere so this is also an opportunity for me to make sure it's got the goods. (so not worried about sunk cost if it doesn't fly.)

I agree it could be sizable, but I also think a lot of groundwork has been laid in that the content already exists for a 1:1 transition.

Re: API visibility, I'd say when in doubt, we exclude it -- can always be included in the docs later.

@ChrisMaddock if you have the insight into the group, can you arrange the necessary conversation / process pieces to move this forward?

**Comment by CharliePoole** [comment]
@SeanKilleen 

Responding to your questions....

 * Yes, I'd say this worked when it worked. Lots of newcomers edited the wiki. It required somebody to curate the contributions (and the contributors) and quickly revert anything that's wrong.

 * Anecdotally, it seems that more and more of those who use NUnit are __not__ devs but testers. They are by no means in the majority but they seem to have a requirement for much more precise and detailed documentation than was needed in the past.

I have a question too...

You cited Akka.Net as a model. I love their docs. It seems clear that a lot of work went into them. Oddly, when I clicked on "Improve this document" I was put into a markdown editor. Of course, I didn't go any further, so I don't know what would happen if I saved a change. But on the surface, it seems like this does exactly what I was saying - no PR involved at all. Am I missing something?

**Comment by SeanKilleen** [comment]
@CharliePoole 

> Lots of newcomers edited the wiki. 

Ah, I see. I'm likely just flat out wrong then 👍  I wasn't anticipating that large of an audience that edited the wiki. 

> It required somebody to curate the contributions (and the contributors) and quickly revert anything that's wrong.

This part is still doable with a PR structure, but I can see how it might have been achievable with a public wiki model too. 

> Anecdotally, it seems that more and more of those who use NUnit are not devs but testers. They are by no means in the majority but they seem to have a requirement for much more precise and detailed documentation than was needed in the past.

One of the bonuses of programmatic documentation in this case then I think is that you could add a button to raise an issue about the doc as well, which could go to a GitHub issues template which helps you understand the context as well. 

> When I clicked on "Improve this document" I was put into a markdown editor. Of course, I didn't go any further, so I don't know what would happen if I saved a change. But on the surface, it seems like this does exactly what I was saying - no PR involved at all. Am I missing something

This is something I like about the new PR workflow. The way that process works in that case is:

* Someone clicks the link
* It links to the markdown file in github
* Someone can click edit, which will make a fork automatically
* Someone can make their suggested changes (in a markdown text editor in the browser, with preview/diff) and then click save
* When they click save, they can then create a PR.

So it does require a GitHub account, but for a lot of docs updates I suggest, I don't use the browser at all.

I follow the same process on my blog; any post has an "edit this post" link, which does the same thing -- takes the user to the markdown file on GitHub which they could then choose to edit and submit a PR (and I do get them coming in).

**Comment by CharliePoole** [comment]
Oh! Well, as Gilda used to say, "Never mind!" So long as naive users don't actually have to clone the repo, etc. we're talking about the same thing. And the direct ediiting process required an account as well. (There was a period of time when publicly edited repos with no security at all worked, but that time passed quite a while back.) :cry:

**Comment by ChrisMaddock** [comment]
> @ChrisMaddock if you have the insight into the group, can you arrange the necessary conversation / process pieces to move this forward?

Happy to tag people in - you'll appreciate some people are a little busy in the current situation though! 🙂 

Here's a quick breakdown of who heads up each what - based on the docs structure.

| Docs area | Main repo | Owner
| ------------- | ------------- |
| "NUnit" | nunit/nunit | @rprouse 
| | nunit/nunit-console  | @ChrisMaddock 
| NUnit Xamarin Runners | nunit/nunit-xamarin | @rprouse 
| Test Adapter | nunit/nunit3-vs-adapter | @OsirisTerje 
| Test Generator | nunit/vs-test-generator | @OsirisTerje 
| NUnit Analyzers | nunit/nunit-analyzers | @mikkelbu 
| NUnit GUI/ NUnit Project Editor | No longer hosted within the organisation | @CharliePoole 

Overall - this probably really wants a look in from the "Core Team", who head up cross-project type decisions. That's @rprouse, @CharliePoole, @OsirisTerje, @jnm2, @mikkelbu and myself. :-)

**Comment by mikkelbu** [comment]
For what it's worth. I just took a look at the last 200 changes to the wiki (meaning back to Nov 13 2017). And if I filter the people mentioned in the previous post away, then I end up with the following contributions:
* Lorenzo Solano Martinez	Wed Jun 12 16:47:12 2019 -0400
* Wheeler Weise	Thu Jul 19 20:19:13 2018 -0500
* Josh Darnell	Thu Feb 8 16:08:56 2018 -0500

So I don't mind mixing thing up to get more help from the outside.

**Comment by CharliePoole** [comment]
I have two concerns, actually:

1. Make it easier for folks to contribute. A process of cloning the wiki repo and updating that way is too complicated for users who don't want to be code contributors. It's more complicated for us as well, since you can't do PRs against the wiki repo. I was just sharing my experience from a few years (i.e. > 10) back, which others here have not had... that is, if you give out edit permissions promiscuously, then some of those people will contribute usefully. @SeanKilleen 's proposal takes that to an extreme by giving __everybody__ some degree of permission. Not sure how merging works... do we review these changes or do can we allow some of them to just update directly?

2. Encourage __us__ to update the docs when needed. That's probably more important. We make changes that require docs and - like programmers everywhere - we sometimes postpone doing it. I'm doubtful that any technical change will make a large difference. I would like to see something that puts the problem more in our faces, like marking docs as potentially out of date when a related code change occurs. 

"When I have the time" I'm going to try this stuff out myself. :smile_cat:

**Comment by SeanKilleen** [comment]
@CharliePoole great points. 

I think it's helpful to see the process first-hand, which is why I started https://github.com/nunit/nunit/pull/3527.

Let's walk through how this process would work using my in-progress PR, say for a common scenario such as a typo. (it would be a similar process whether using docfx or any other static site, such as my blog, that is able to link back to the markdown files). 

I head to the page for `Assert.IsFalse`, where I see a(n intentionally-inserted) typo: 

> ![image](https://user-images.githubusercontent.com/2148318/80521023-51292600-8958-11ea-9ef1-6433c2e7a6d9.png)

I see the link on the right-hand side to improve the doc: 

> ![image](https://user-images.githubusercontent.com/2148318/80521056-61410580-8958-11ea-9a20-bf66997b34b9.png)

I click the link, and I am taken to the markdown file within GitHub -- with an edit button:

> ![image](https://user-images.githubusercontent.com/2148318/80521171-92b9d100-8958-11ea-8915-eaa11069cb7c.png)

I click the edit button (yes, this does require me to have a GitHub account, which is unfortunate.)

I fix the typo:

> ![image](https://user-images.githubusercontent.com/2148318/80521252-b7ae4400-8958-11ea-868b-701e6028eaac.png)

I preview the changes:

> ![image](https://user-images.githubusercontent.com/2148318/80521279-c694f680-8958-11ea-9a94-e9037c22cd14.png)

I now am on to the part where I can request my changes. Things to note:

* GitHub auto-populates a commit message, which while not the most descriptive, is fine to get things started.
* It automatically suggests a default branch name (it will automatically create a fork if needed)
* I can choose to add a description or not
* There's a link for helpful guidance if someone wants to know more.

> ![image](https://user-images.githubusercontent.com/2148318/80521570-40c57b00-8959-11ea-83b8-fcfa9bb8d805.png)

I click to propose the file change by clicking the button. I now see a pull request populated with my suggested changes.

> ![image](https://user-images.githubusercontent.com/2148318/80522581-ce559a80-895a-11ea-969c-57f8c8de455c.png)

So: 

| Pros | Cons | 
| ----- | ------|
| ✅ Easy to link from the site | ❌ GitHub Login still required |
| ✅ Never need to know a git command | ❌ May not be intuitive (less than wiki process? unsure) |
| ✅ No manual forking/cloning necessary | |
| ✅ Known accepted process for making changes; thus invites them | |
| ✅ Paradigm shared in many other projects docs, etc. | |

**Comment by CharliePoole** [comment]
Thanks Sean. I knew most of that but the example is very helpful. I assume that the "commit directly" radio button depends on the particular user having write access. If they don't, then it all comes down to what processes will be implemented to merge the changes.

I think this is completely independent of what you are doing and is driven by the team, particularly by the lead of the website project. But currently, I have no authority to check the button that leads to a direct commit. That's what drives my comments.

In comparison, just this morning I did an immediate update of a web site page in response to a clear misunderstanding it was causing a user. The difference there isn't the technology, which allowed me to do it, it's in the surrounding policies which gave me permission.

What I wish we could do is (1) adopt this tech you are proposing, (2) eliminate all gatekeeping (PRs) for org members unless they elect to ask for a review and (3) do the same thing for outside contributors by inviting them inside in some sort of editor role.

**Comment by SeanKilleen** [comment]
@CharliePoole 

> What I wish we could do is (1) adopt this tech you are proposing, (2) eliminate all gatekeeping (PRs) for org members unless they elect to ask for a review and (3) do the same thing for outside contributors by inviting them inside in some sort of editor role.

That is doable in a few ways. 

* Anyone who has access to push code also has access to push docs, no PRs needed. 
* You could put the docs in a separate repo in order to allow more permissive push access to the master branch (you could use another branch in a given repo too, but that reduces visibility so I don't like this option)
* You could create a group within the org, e.g. `docs-contributors`, add folks, and then exempt them from any PR review checks.
* This could also be done by a bot that auto-approves a merge based on the users in a group.  

Might take a little fiddling with the details, but what you've described is achievable. You just don't get it with docfx alone.

**Comment by CharliePoole** [comment]
@SeanKilleen Exactly! "You don't just get it with docfx alone" is the point I'm trying to make to my fellow core team members. 

@nunit/core-team I'm afraid that @SeanKilleen has been caught in the crossfire. I think that dotfx is a great idea - so great I'm planning to use it for my own docs. But without opening up the process so that changes either don't require approval or at least get approved more rapidly I'm dubious about the results we will see out of it.

Currently, according to the governance docs, nunit.org is a secondary project. It doesn't have a Team, only a lead, whom I believe is still @rprouse. Only the lead has authority to commit anything. (Again, not talking GitHub authority but our governance "rules" as they were set up) 

We set up our governance docs so that each "project" could function as it saw fit, without any control from the Core Team. I'm wondering now if the documentation project and possibly the websiteas a whole, should be handled in a more common way and perhaps have a larger number of people who are able to commit to it.

**Comment by rprouse** [comment]
Wow, that is a lot of discussion in 4 days to catch up on! I am very much in favor of switching our docs away from the wiki format. We thought that the wiki would encourage less technical developers to easily contribute to our docs, but that has not been the case. In the past, we've added people to the docs team, they've made a change, then moved on. None of the docs team has stuck around. We had the best intentions, but I think we need to re-evaluate.

I think we need to be realistic. The average person using NUnit these days who would be willing to contribute to our documentation is probably a developer who can handle creating a pull request and who can definately handle the web based process that @SeanKilleen documented above.

I think that we only need to look at the amazing job that the Microsoft Docs team has done for proof that the process can work. I've contributed to the Microsoft docs and the PR process is very easy and straight-forward.

We do have a number of projects that need to be considered, each with slightly different needs. The core is the NUnit Framework docs themselves and I think they are most important as they are the information on how to write unit tests with NUnit. If we are going to start anywhere, I'd like to start here.

The Console/Engine project is a bit different. It covers how to run tests from the command line. It should also cover more about how to use and extend the engine, but we haven't done a great job documenting that.

We then have the other projects like the VS Adapter which is seeing more and more use and needs better documentation. There is also the Analyzer project and the templates. These are smaller, but will need expanding over time.

@SeanKilleen I like what you are doing in your PR. I like that the docs are with the code as we could require that code changes include the corresponding documentation changes in the PR the same way we require tests.

How will this work when we start documenting the other projects?

Anyway, TLDR, I like where this is going. Let's see where this leads.

**Comment by ChrisMaddock** [comment]
I'd be keen to extract and expand the Engine & Console docs in parallel with Sean's work here. 🙂 

Albeit when I have more time, as always! 😄

**Comment by ChrisMaddock** [comment]
> What I wish we could do is (1) adopt this tech you are proposing, (2) eliminate all gatekeeping (PRs) for org members unless they elect to ask for a review and (3) do the same thing for outside contributors by inviting them inside in some sort of editor role.

No objections here from me. It reflects how we've been treating the docs anyway. Only thing that might make this more complex is if we move the docs into each code repo (as Rob suggests above). Pro's and con's - helps with maintenance, but means permissions should reflect code permissions. I think the pros outweigh the cons here - anything to help us keep the docs better up to date!

**Comment by jnm2** [comment]
I'm surprised and grateful that someone is willing to handle what at least I've considered an overwhelming task and an area of some debt. Thanks @SeanKilleen!

On whether changes should be reviewed, I'd prefer a required PR process even for core members. I like the paradigm of treating docs with the same care as code. From a contributor's perspective, it's not inherently any more effort to submit a change for review than to have the changes take effect immediately, unless the reviewers find the change unacceptable as-is. So either the contributor never has to come back to it after first submitting the change, or we catch mistaken information before it goes public as from us.

**Comment by SeanKilleen** [comment]
Thanks for the great discussion here, all! An update on https://github.com/nunit/nunit/pull/3527: 

* I have the documentation built for NUnit (a few more articles to port from missing links, etc. and some formatting, but the content is 95% there).
* I have no content built for the other projects yet (e.g. vs test adapters, etc.) 

I see a few potential paths forward, and there may be others I haven't considered:

1. Move the docs to the `nunit/docs` repository for all docs across the org
1. Move the docs to the `nunit/nunit` repo (where my PR is now)
1. Create docs in a `docs` folder within each individual project, and cross-reference them as different external documentation sites (harder to host, but keeps autonomy clear).

Given my current understanding, my recommendation would be "one docs site to rule them all" that lives in the `nunit/docs` repo, because it's a clear place for people to find and contribute. 

Could this group please help me understand what the consensus is on the project maintainers so that I can adjust accordingly and open a new PR if necessary?

If you choose to publish the API documentation, I believe there are still ways to do that across projects (a build process for docs that can generate the metadata for external repos, for example).

I'd love to understand a consensus from the team so that I can keep things moving forward; want to make sure my momentum doesn't die down.

**Comment by CharliePoole** [comment]
@nunit/core-team My thought is that this is something for us to discuss... i.e. whether all projects will continue to use common docs vs having them in each project. Or do you feel this is a decision we already made when we put the docs in a common repo?

The other common place we have is in the website. @SeanKilleen Do you think it's also an option to use a subdomain in our website? Or does the fact that it's built using Jekyll make that overly complicated? (I'm not advocating we do that, just trying to list all options)

**Comment by SeanKilleen** [comment]
@CharliePoole 

> Do you think it's also an option to use a subdomain in our website? 

Should be. This is a web site that is generated and will need to be hosted somewhere. GitHub pages, Netlify, azure, AWS all allow static site hosting for custom domains. So once you pick a thing to run it, it would be a matter of configuring a subdomain within your domains DNS settings to point to that provider. I'd be happy to help achieve that via a Zoom call, etc.

> Or does the fact that it's built using Jekyll make that overly complicated? (I'm not advocating we do that, just trying to list all options)

If you're referring to the nunit site being built in Jekyll, it opens up another option -- we could move this doc site from docfx to Jekyll and combine the two, using Jekyll for the docs as well. In this case I'd recommend keeping them separate since they have separate purposes and docfx shines for docs. However I've got experience in Jekyll and happy to help port there as well.

**Comment by CharliePoole** [comment]
I agree with you about docfx being better for documentation. That would especially be true if we wanted to phase out some doc pages that are hand-written in favor of generating them from the code.

**Comment by SeanKilleen** [comment]
Sounds good.

My current intention is to fix up the content I have so that it's representative of what the final state would look like. I'll keep making commits on the PR to that effect.

In the interim this group can decide:

- whether they want the additional content (e.g. vs test runner) in these docs 
- where they would ultimately like all the docs to live.


If you'd like to see the current state of things, I'll be adding an automated build soon but for now you can:

-  install docfx (I recommend chocolatey: `choco install docfx`)
- check out the or
- `cd docs`
- `docfx metadata`
- `docfx build ; docfx serve`

**Comment by CharliePoole** [comment]
@SeanKilleen Nitpicky clarification... Assuming "this group" means the framework team, I believe that the Core team is actually who has to decide. Important to me and one other because we are on the Core team but not the framework team.

This wouldn't be a big deal to me, in fact, except it came up as an individual project issue even though we are at least talking about other projects. I don't think a decision that could affect other projects substantially ought to be made without full involvement, which is what the Core Team is designed to give us.

**Comment by SeanKilleen** [comment]
@CharliePoole what I'm saying is, as someone outside the project, I have no idea who the decider is :) but I'd like someone on the project to volunteer to get the right people together to discuss and make a decision. 👍

**Comment by CharliePoole** [comment]
@rprouse 

I realize you're pretty busy, but Sean is asking for a volunteer to help move this forward and it seems to me that it's entirely in your court. So I am volunteering to tell you that. :smile: 

You're the lead of both the groups potentially involved: the Framework Team and the Core Team. If you wanted separate docs for the framework project, as the PR is now set up, then I think that's what would happen. In a Core Team vote, I'd support separation if even one project lead wanted them separate and I bet most people would do the same.

OTOH, if you think this is a Core Team decision, as I do, our governance documents give you __and nobody else__ the power to call for a meeting and/or discussion. (Frankly, I'd like to change that, but that's another matter.)

@SeanKilleen Let me repeat that I appreciate your work especially because it has potential to be used beyond the framework project. I'm even planning to use `docfx` in my external projects, thanks to your presentation of what it can do.

**Comment by ChrisMaddock** [comment]
I have no strong feelings either way. In the same repo as the code has the advantage of it being easier to ask committers to document their changes all in one PR. 😊

**Comment by ChrisMaddock** [comment]
> whether they want the additional content (e.g. vs test runner) in these docs

Afraid I’m not totally clear what you’re asking with this point Sean - but I think we should still have one seamless docs site for all projects, regardless of where the markdown to build them from lives. 😊

**Comment by SeanKilleen** [comment]
@ChrisMaddock 

> Afraid I’m not totally clear what you’re asking with this point Sean - but I think we should still have one seamless docs site for all projects, regardless of where the markdown to build them from lives

I was referring to whether we want to port the wikis from the Xamarin runner, vs test adapters, analyzers, etc. into this same site.

Upon further inspection, it looks like these are already topics under the same wiki, so I will port them as well -- no need for a decision on that point. 👍

**Comment by rprouse** [comment]
@SeanKilleen are you planning on pulling all of the documentation for all of the projects that are currently hosted in this repo's Wiki and pulling them into your current PR, or are you planning on separate PRs and doc sites per project? As @CharliePoole states, if we are going to separate the documentation for the projects, we should probably check with the other project leads and the core team first, but your statements lead me to believe that you are pulling all of the current wiki docs in eventually and that the framework docs are just the first step.

Can you clarify please?

I'd like to reiterate that all of us really appreciate what you are doing and I think it is a big move in the right direction. Please bear with us on this one, it is a big change for the project and we want to make sure we get it right. Thanks.

**Comment by SeanKilleen** [comment]
@rprouse no worries at all! I'm happy to take direction from the group either way; my goal is to move this effort forward in whatever manner helps the group the most. Nothing wrong with smaller chunks of work.

Multiple PRs are fine with me; I am looking for guidance on where to demarcate the content that I'm pulling in.

Originally, I was planning on leaving the below topics from [the wiki nav](https://github.com/nunit/docs/wiki/NUnit-Documentation) out -- as this conversation progressed, I began pulling them in (but could easily separate them out again with very little effort).

> ![image](https://user-images.githubusercontent.com/2148318/81131318-c4163c00-8f18-11ea-999e-a4a9667dc2ff.png)

If you'd like me to de-scope them for now, I can create an issue to track the overall work, and scope my first PR to the NUnit Framework docs.

**Comment by jnm2** [comment]
Has anyone mentioned a downside to having all NUnit projects coordinate on docs through a single repo such as nunit/docs?) It sounds less busy-work all around.

**Comment by ChrisMaddock** [comment]
In my mind, the main advantage we would get of moving docs into code-repo's would be that it's easier for a contributor to update docs and code in a single PR - something which we're not too great at doing at the moment.

That said - it brings a level of complexity to the setting-up of docfx. If I was doing this, I'd be tempted to keep everything in nunit/docs for the initial set up, and then consider whether we want to move stuff out to the different project's repos as phase two?

> If you'd like me to de-scope them for now, I can create an issue to track the overall work, and scope my first PR to the NUnit Framework docs.

My view, which others may disagree with(!) is that there's two acceptable solutions here:

1. Keep all the docs in nunit/docs
2. Having docfx pull in different docs from the 7(?) different project repos

Option 1 is of course simpler in terms of set up! 🙂 The options I'd be less happy with:

3. Setting up a docfx site for just the framework docs, and leaving the other docs in the current GitHub wiki. I think having one site for all NUnit docs is a good thing - users don't always see the same distinction between the projects that the team do.
4. Moving all the docs into the nunit/nunit repo. It would be strange to have documentation for a 'different project' bundled up in the framework repo.

**Comment by CharliePoole** [comment]
Starting with the low-hanging fruit...

I absolutely agree with @ChrisMaddock  on options 3 and 4.

Between 1 and 2, I like keeping the documentation source in each project, while having the docs all be in one place. @SeanKilleen indicated this was an option. I don't know what that one place should be but since we are building static web pages, I guess it belongs on some website and nunit.org is the one we've got. Maybe if we agreed on the end goal, @SeanKilleen could tell us how it might be done. One thought, born out of my complete lack of knowledge of how this works, is to have the source be in each project, the generated html be combined under docs and the web site to reach out and serve it from there.

One thing we haven't yet talked about is how to implement this. There is a problem where the minute we grab a page from the wiki and move it to the repo, that page can still be changed at any moment. I know I have chaned a page or two since Sean started. So it seems like it's going to require a final check to make sure we have everything and that it would serve us well to do as small a bite as is reasonable at one time. Since it would inconvenience users to do less than a full project, that would seem like the proper unit of migration. What do others think?

**Comment by rprouse** [comment]
Either of @ChrisMaddock's first two options sound good to me and I agree with @CharliePoole that we should do at least one full project at a time and I think starting with one like the framework would be a good way to go about it.

>There is a problem where the minute we grab a page from the wiki and move it to the repo, that page can still be changed at any moment. I know I have chaned a page or two since Sean started. So it seems like it's going to require a final check to make sure we have everything and that it would serve us well.

That is a good idea. We can do the check by cloning the Wiki repo and checking the history.

**Comment by jnm2** [comment]
It sounds promising to make it standard in every PR to consider whether to update docs in the same branch. I like that concept.

**Comment by mikkelbu** [comment]
I can also go with either of the first two options from @ChrisMaddock. Currently, the analyzer documentation on the wiki is just copy-paste of the markdown files from the analyzer repo except for a subsequent search-replace to change markdown links to wiki links.

**Comment by SeanKilleen** [comment]
> If I was doing this, I'd be tempted to keep everything in nunit/docs for the initial set up, and then consider whether we want to move stuff out to the different project's repos as phase two?

I agree with @ChrisMaddock here. My recommendation to get things rolling would be to have the docs in one repo, `nunit/docs`. I would suggest serving this content at `docs.nunit.org` or similar. 

In the interim, code owners can be specified for the markdown files within github for certain sections of the docs, which will help you demarcate between projects should you desire that.

Long-term, it may make sense to keep these docs with the individual repos, but I think that it's a big over-complication initially. If you launched `docs.nunit.org` with docfx from the `nunit/docs` repo, it shouldn't be a big lift later to separate them out if it's truly desired. 

I think the gains you'll get initially from doing it in one repo will be enough without having it side-by-side in source code. It's one spot to go for anyone who wants information or to contribute, and it's easy enough to track needed updates to the docs as related issues when making code changes. 

Docfx is not designed by default to generate for multiple repos. You'd likely end up having each repo with a docfx.json file, and stitching them together as folder contents via a build process. Just adds a bit more moving parts than would be helpful right now, IMO.

**Comment by SeanKilleen** [comment]
Also, to @CharliePoole's concern:

> There is a problem where the minute we grab a page from the wiki and move it to the repo, that page can still be changed at any moment.

In the interim between the PR being merged and the `docs.nunit.org` site being published, I would suggest tracking docs changes as issues in `nunit/docs` to be applied later.

In my current PR, I have a checklist item to check newer docs.

The NUnit github wiki can be downloaded as a repo, so I have a change tracker with every change since I've begun this process, and I can ensure the content is up to date (and I intend to.)

**Comment by CharliePoole** [comment]
@SeanKilleen Something I have never been clear on... do changes made through the browser interface to the wiki (just clicking edit) get into that history? If they don't, maybe we should shut down the edit interface temporarily.

**Comment by SeanKilleen** [comment]
@CharliePoole I believe so but am not certain. If anyone wants to make a small change via the browser now, I could quickly verify.

**Comment by CharliePoole** [comment]
Just a moment. :smile:

**Comment by CharliePoole** [comment]
Yup. Verified by editing the markdown for my GUI home page online and then pulling.

**Comment by ChrisMaddock** [comment]
Hey Sean,

Looks like we have agreement from 5/6 of the team that keeping the content in nunit/docs is the way to go for now, so I think we're good to go with that. (80% is a high number for us!) 

@CharliePoole - I've assumed from you making no further comment on [Sean's response](https://github.com/nunit/docs/issues/337#issuecomment-626242304) that you're happy with this as a way forward?

In terms of hosting and subdomains, it's @rprouse who holds the relevant accounts there. Conceptually, I think docs.nunit.org sounds great, and hosting on GitHub pages seems to make the most sense, to be consistent with [the organisations website](https://github.com/nunit/nunit.github.io).

**Comment by CharliePoole** [comment]
@ChrisMaddock  

I hadn't seen Sean's comment, but I can go with that. I would be interested to see how it works to have them there if there is some content which we generate from the code rather than authoring text documents. Large portions of the nunit docs as well as most of the engine API could potentially be handled that way.

**Comment by SeanKilleen** [comment]
FYI all, I've opened #339 to move these changes into the docs repo and closed https://github.com/nunit/nunit/pull/3527. Will continue the effort there.

**Comment by SeanKilleen** [comment]
With http://docs.nunit.org, I am happy to be able to close this issue. Thanks for putting some faith in me to get it done. 

Onward! 🚀
